pallets/remarks/src/tests.rs

```diff
@@ -0,0 +1,228 @@
+				remark: vec![3, 4, 5],
+			});
+
+			let expected_pre_origin = origin.clone();
+			let expected_pre_remarks = remarks.clone();
+			let expected_pre_call = call.clone();
+
+			RemarkDispatchHandlerMock::mock_pre_dispatch_check(move |t| {
+				assert_eq!(t.0.as_signed(), expected_pre_origin.clone().as_signed());
+				assert_eq!(t.1, expected_pre_remarks);
+				assert_eq!(t.2, Box::new(expected_pre_call.clone()));
+
+				Ok(())
+			});
+
+			let expected_post_origin = origin.clone();
+			let expected_post_remarks = remarks.clone();
+			let expected_post_call = call.clone();
+
+			RemarkDispatchHandlerMock::mock_post_dispatch_check(move |t| {
+				assert_eq!(t.0.as_signed(), expected_post_origin.clone().as_signed());
+				assert_eq!(t.1, expected_post_remarks);
+				assert_eq!(t.2, Box::new(expected_post_call.clone()));
+
+				Ok(())
+			});
+
+			assert_ok!(Remarks::remark(
+				origin.clone(),
+				remarks.clone(),
+				call.clone().into()
+			));
+
+			let expected_event: RuntimeEvent = Event::<Runtime>::Remark { remarks, call }.into();
+
+			System::events()
+				.iter()
+				.find(|e| e.event == expected_event)
+				.expect("remark event is present");
+		});
+	}
+
+	#[test]
+	fn no_remarks() {
+		new_test_ext().execute_with(|| {
+			let call = RuntimeCall::System(SystemCall::remark {
+				remark: vec![3, 4, 5],
+			});
+
+			assert_noop!(
+				Remarks::remark(
+					RuntimeOrigin::signed(1),
+					Default::default(),
+					call.clone().into()
+				),
+				Error::<Runtime>::NoRemarks,
+			);
+		});
+	}
+
+	#[test]
+	fn pre_dispatch_failure() {
+		new_test_ext().execute_with(|| {
+			let remarks = get_test_remarks();
+
+			let call = RuntimeCall::System(SystemCall::remark {
+				remark: vec![3, 4, 5],
+			});
+
+			let expected_error = DispatchError::Other("pre-dispatch error");
+
+			RemarkDispatchHandlerMock::mock_pre_dispatch_check(move |_t| Err(expected_error));
+
+			assert_noop!(
+				Remarks::remark(
+					RuntimeOrigin::signed(1),
+					remarks.clone(),
+					call.clone().into()
+				),
+				expected_error
+			);
+		});
+	}
+
+	#[test]
+	fn post_dispatch_failure() {
+		new_test_ext().execute_with(|| {
+			let remarks = get_test_remarks();
+
+			let call = RuntimeCall::System(SystemCall::remark {
+				remark: vec![3, 4, 5],
+			});
+
+			RemarkDispatchHandlerMock::mock_pre_dispatch_check(move |_t| Ok(()));
+
+			let expected_error = DispatchError::Other("post-dispatch error");
+
+			RemarkDispatchHandlerMock::mock_post_dispatch_check(move |_t| Err(expected_error));
+
+			assert_noop!(
+				Remarks::remark(
+					RuntimeOrigin::signed(1),
+					remarks.clone(),
+					call.clone().into()
+				),
+				expected_error
+			);
+		});
+	}
+
+	#[test]
+	fn nested_remark_call_failure() {
+		new_test_ext().execute_with(|| {
+			let remarks = get_test_remarks();
+
+			let call = RuntimeCall::Remarks(Call::remark {
+				remarks: Default::default(),
+				call: Box::new(RuntimeCall::System(SystemCall::remark {
+					remark: vec![3, 4, 5],
+				})),
```

Comment:
Do we want to add a similar test which nests `Remark::remark` inside each other?

Reply:
We already have that in `nested_remark_call_failure`, or do you mean some other kind of nesting?

Reply:
As discussed, we already have that in. I also added just now another test for proxy calls to be extra sure though.

Reply:
Thanks for adhering to my paranoia 😅